gemini/orchestration/autogen_gemini.ipynb

```diff
@@ -0,0 +1,603 @@
+        "    query: Annotated[\n",
+        "        str, \"A natural language or structured query containing a location\"\n",
+        "    ]\n",
+        ") -> tuple[float, float] | None:\n",
+        "    \"\"\"Performs a free-form location search using the Nominatim API.\"\"\"\n",
+        "    base_url = \"https://nominatim.openstreetmap.org/search\"\n",
+        "    params = {\n",
+        "        \"q\": query,\n",
+        "        \"format\": \"jsonv2\",\n",
+        "        \"addressdetails\": \"1\",\n",
+        "        \"email\": \"your_email@example.com\",  # Replace with your email\n",
+        "    }\n",
+        "    headers = {\n",
+        "        \"User-Agent\": f\"MyWeatherApp/1.0 ({params['email']})\",\n",
+        "    }\n",
+        "    try:\n",
+        "        response = requests.get(base_url, params=params, headers=headers)\n",
+        "        response.raise_for_status()\n",
+        "        search_results = response.json()\n",
+        "        lat = float(search_results[0][\"lat\"])\n",
+        "        lon = float(search_results[0][\"lon\"])\n",
+        "        return lat, lon\n",
+        "    except requests.exceptions.RequestException as e:\n",
+        "        print(f\"Error during Nominatim API request: {e}\")\n",
+        "        return None\n",
+        "    finally:\n",
+        "        time.sleep(1)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 7,
+      "metadata": {
+        "id": "kv1hYDB76vG7"
+      },
+      "outputs": [],
+      "source": [
+        "@user_proxy.register_for_execution()\n",
+        "@weather_agent.register_for_llm(\n",
+        "    description=\"Retrieves the weather forecast for a given latitude and longitude.\"\n",
+        ")\n",
+        "def get_weather_forecast(\n",
+        "    latitude: Annotated[float, \"Distance north or south of the equator\"],\n",
+        "    longitude: Annotated[float, \"Distance east or west of the prime meridian\"],\n",
+        ") -> dict | None:\n",
+        "    \"\"\"Retrieves the weather forecast using the Open Meteo API.\"\"\"\n",
+        "    base_url = \"https://api.open-meteo.com/v1/forecast\"\n",
+        "    params = {\n",
+        "        \"latitude\": str(latitude),\n",
+        "        \"longitude\": str(longitude),\n",
+        "        \"daily\": \"temperature_2m_max,temperature_2m_min,precipitation_sum,windspeed_10m_max\",\n",
+        "        \"timezone\": \"auto\",\n",
+        "    }\n",
+        "    try:\n",
+        "        response = requests.get(base_url, params=params)\n",
+        "        response.raise_for_status()\n",
+        "        weather_forecast = response.json()\n",
+        "        print(weather_forecast)\n",
+        "        return weather_forecast\n",
+        "    except requests.exceptions.RequestException as e:\n",
+        "        print(f\"Error during Open Meteo API request: {e}\")\n",
+        "        return None"
```

Comment:
Similar to above, could you try using the Python Library for Open Meteo?

https://pypi.org/project/open-meteo/

Reply:
Talked offline, will resolve later